src/CacheWithResilience.php

```diff
@@ -0,0 +1,165 @@
+        CacheItemPoolInterface $cachePool,
+        string $prefix,
+        array $cacheTimes = [],
+        array $whitelistedExceptions = []
+    ) {
+        $this->logger = $logger;
+        $this->whitelistedExceptions = $whitelistedExceptions;
+        parent::__construct($cachePool, $prefix, $cacheTimes);
+    }
+
+    /**
+     * This function get the item if exist in memory and the TTL
+     *
+     * @param string $key
+     * @param bool $returnStaleValue
+     * @return CacheItemInterface
+     */
+    public function getItem(string $key, bool $returnStaleValue = false): CacheItemInterface
+    {
+        $key = $this->standardiseKey($key);
+        if ($this->flushCacheItems) {
+            $this->cachePool->deleteItem($key);
+        }
+
+        $item = $this->cachePool->getItem($key);
+        if ($item->isHit()) {
+            $value = $item->get();
+            // we have 2 TTL's, one of them is stored with the item as $value['ttl'] and is the time
+            // when we want the item to be consider as expired, the second one is the real TTL that we tell redis.
+            // The second one is only mean to be used in case of issues like DB down and only requires to
+            // set $returnStaleValue to true
+            if (false === $returnStaleValue && new DateTime() > $value['ttl']) {
+                return new CacheItem();
+            }
+        }
+
+        // in setItem() value gets wrapped into an array with the ttl. set the item to its original value
+        $val = $item->get();
+        $item->set($val['value']);
+
+        return $item;
+    }
+
+    /**
+     * @param CacheItemInterface $item
+     * @param mixed $value
+     * @param int|string|DateTimeInterface $ttl
+     *   TTL in seconds, or a constant from CacheInterface, or a DateTime to expire at
+     * @return bool
+     */
+    public function setItem(CacheItemInterface $item, $value, $ttl): bool
+    {
+        // this condition calculates the DateTime when the element should be consider as expire
+        if ($ttl instanceof DateTimeInterface) {
+            $expiresAt = $ttl;
+        } elseif (\in_array($ttl, [0, -1, CacheInterface::NONE, CacheInterface::INDEFINITE, CacheInterface::SHORT], true)) {
+            // this case handles a list of exception that shouldn't be cached for a different length
+            $ttl = $this->calculateTtl($ttl);
+            $item->expiresAfter($ttl);
+            // getItem() expects the item to be wrapped as [value, ttl]
+            $item->set(['value' => $value, 'ttl' => $ttl]);
+            return $this->cachePool->save($item);
+        } else {
+            $ttl = $this->calculateTtl($ttl);
+            $expiresAt = new DateTime();
+            $expiresAt = $expiresAt->add(new DateInterval('PT' . $ttl . 'S'));
+        }
+
+        // save the ttl within the stored item so we can know if a returned key is expired.
+        $item->set(['value' => $value, 'ttl' => $expiresAt]);
```

Comment:
"time to live" (seconds) now holds an "expiry date" (date)

Reply:
probably I should name it `expiresAt` as well

Reply:
either way works, as long as the key name reflects the type